<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2023-08-13" />

<title>DE Annotations &amp; Downstream applications</title>

<script src="site_libs/header-attrs-2.18/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02a_Sequence_QC.html">Sequence QC</a>
    </li>
    <li>
      <a href="Module02b_Trimming.html">Trimming</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03a_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module03b_Alignment.html">Alignment</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module04_Alignment_QC_and_Quantification.html">Alignment QC and Quantification</a>
    </li>
    <li>
      <a href="Module05_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">DE Introduction</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DE Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">DE Normalization and Modeling</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Testing</a>
    </li>
    <li>
      <a href="Module11_DEVisualizations.html">DE Visualization</a>
    </li>
    <li>
      <a href="Module12_DEAnnotations.html">DE Annotation</a>
    </li>
    <li>
      <a href="R_bonus_content.html">Bonus content</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">DE Annotations &amp; Downstream
applications</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2023-08-13</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<p>In this module, we will learn:</p>
<ul>
<li>The advantages of using gene ids when analyzing RNA-seq data.</li>
<li>How to add gene symbols and other annotations, using ENSEMBL gene
ids</li>
<li>Discuss options for functional enrichments and other follow-ups</li>
</ul>
<p><br></p>
<div id="differential-expression-workflow"
class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Differential Expression Workflow</h1>
<p>Here we will generate summary figures for our results and annotate
our DE tables.</p>
<p><img src="images/wayfinder/slides/Slide12.png"
style="width:75.0%" /></p>
<hr />
</div>
<div id="adding-genome-annotations" class="section level1">
<h1>Adding genome annotations</h1>
<p>Since, gene symbols can change over time or be ambiguous we use, and
recommend, using the EMSEMBL reference genome and ENSEMBL IDs for
alignments and we’ve been working with tables and data where all genes
are labeled only by their long ENSEMBL ID. However, this can make it
difficult to quickly look for genes of interest.</p>
<p>Luckily, Bioconductor provides many tools and resources to facilitate
access to <a
href="http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html">genomic
annotation resources</a>.</p>
<p>To start, we will first load the <a
href="https://bioconductor.org/packages/3.14/bioc/html/biomaRt.html">biomaRt
library</a> and choose what reference we want to access. For a more
detailed walk through of using biomaRt, <a
href="https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2019/RNAseq/html/05_Annotation_and_Visualisation.html">this
training module</a> might be useful, including what to do when
annotations are not 1:1 mappings.</p>
<p>We’ll start by loading the <code>biomaRt</code> library and calling
the <code>useEnsembl()</code> function to select the database we’ll use
to extract the information we need. This will download the mapping of
ENSEMBL IDs to gene symbols, enabling us to eventually add the gene
symbol column we want.</p>
<pre class="r"><code>library(&#39;biomaRt&#39;)
ensembl = useEnsembl(dataset = &#39;mmusculus_gene_ensembl&#39;, biomart=&#39;ensembl&#39;, mirror = &quot;useast&quot;)</code></pre>
<p><strong>Note</strong> <em>- this process takes some time and will
take up a larger amount of working memory so proceed with caution if you
try to run these commands on a laptop with less than 4G of
memory</em></p>
<p>To identify possible <strong>filters</strong> to restrict our data,
we can use the <code>listFilters</code> function. To identify the
<strong>attributes</strong> we want to retrive, we can use the
<code>listAttributes</code> function. The best approach is to use <a
href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html#how-to-build-a-biomart-query">list
or search functions</a> to help narrow down the available options.</p>
<pre class="r"><code>head(listFilters(mart = ensembl), n = 20)
head(listAttributes(ensembl), n = 30)</code></pre>
<p>We can access additional genomic annotations using the <a
href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html"><code>bioMart</code>
package</a>. To identify we’ll structure our ‘query’ or search of the
bioMart resources to use the <a
href="https://m.ensembl.org/info/genome/genebuild/gene_names.html">ENSEMBL
id</a> from our alignment to add the gene symbols and gene description
for each gene.</p>
<pre class="r"><code>id_mapping = getBM(attributes=c(&#39;ensembl_gene_id&#39;, &#39;external_gene_name&#39;),
      filters = &#39;ensembl_gene_id&#39;,
      values = row.names(assay(dds_fitted)),
      mart = ensembl)
# will take some time for query

# Preview the result
head(id_mapping)</code></pre>
<pre><code>     ensembl_gene_id external_gene_name
1 ENSMUSG00000000001              Gnai3
2 ENSMUSG00000000028              Cdc45
3 ENSMUSG00000000031                H19
4 ENSMUSG00000000037              Scml2
5 ENSMUSG00000000049               Apoh
6 ENSMUSG00000000056               Narf</code></pre>
<p>Now that we have the ENSEMBL information and a gene symbol to match
to our results, we can proceed in the smaller groups. As with the
previous exercise, we have broken it into small steps with hints as
needed.</p>
<p><strong>Note</strong>: For additional information regarding bioMart,
please consult the <a
href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html">ENSEMBL
bioMart vignette</a> or the broader <a
href="http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html">Bioconductor
Annotation Resources vignette</a>.</p>
<p>Look at the two data frames that are going to be needed:
<code>id_mapping</code> and
<code>results_deficient_vs_control</code>.</p>
<pre class="r"><code>head(id_mapping)</code></pre>
<pre><code>     ensembl_gene_id external_gene_name
1 ENSMUSG00000000001              Gnai3
2 ENSMUSG00000000028              Cdc45
3 ENSMUSG00000000031                H19
4 ENSMUSG00000000037              Scml2
5 ENSMUSG00000000049               Apoh
6 ENSMUSG00000000056               Narf</code></pre>
<pre class="r"><code>head(results_deficient_vs_control)</code></pre>
<pre><code>log2 fold change (MLE): condition deficient vs control 
Wald test p-value: condition deficient vs control 
DataFrame with 6 rows and 7 columns
                      baseMean log2FoldChange     lfcSE      stat
                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSMUSG00000000001  1489.83039       0.297760  0.210310  1.415815
ENSMUSG00000000028  1748.93544       0.226421  0.176795  1.280695
ENSMUSG00000000031  2151.87715       0.457635  0.764579  0.598545
ENSMUSG00000000037    24.91672       0.579130  0.561259  1.031840
ENSMUSG00000000049     7.78377      -0.899483  1.553063 -0.579167
ENSMUSG00000000056 19653.54030      -0.174048  0.203529 -0.855151
                      pvalue      padj     call
                   &lt;numeric&gt; &lt;numeric&gt; &lt;factor&gt;
ENSMUSG00000000001  0.156830  0.868573       NS
ENSMUSG00000000028  0.200301  0.902900       NS
ENSMUSG00000000031  0.549476  0.995391       NS
ENSMUSG00000000037  0.302147  0.950613       NS
ENSMUSG00000000049  0.562476  0.998043       NS
ENSMUSG00000000056  0.392468  0.982479       NS</code></pre>
<p>We want to match the <code>id</code> column of
<code>results_deficient_vs_control</code> to the
<code>ensembl_gene_id</code> column of <code>id_mapping</code>, and once
that match is found, we want to extract the
<code>external_gene_name</code> column of <code>id_mapping</code> to get
the gene symbol. Next, look at the documentation for
<code>dplyr::left_join()</code> and merge the <code>id_mapping</code>
table into the <code>results_deficient_vs_control</code> table on the
columns <code>ensembl_gene_id</code> and
<code>external_gene_name</code>.</p>
<pre class="r"><code>results_deficient_vs_control_annotated = as_tibble(results_deficient_vs_control, rownames = &quot;id&quot;) %&gt;%
    left_join(id_mapping, by = c(&#39;id&#39; = &#39;ensembl_gene_id&#39;))
head(results_deficient_vs_control_annotated)</code></pre>
<pre><code># A tibble: 6 × 9
  id          baseM…¹ log2F…² lfcSE   stat pvalue  padj call  exter…³
  &lt;chr&gt;         &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;fct&gt; &lt;chr&gt;  
1 ENSMUSG000…  1.49e3   0.298 0.210  1.42   0.157 0.869 NS    Gnai3  
2 ENSMUSG000…  1.75e3   0.226 0.177  1.28   0.200 0.903 NS    Cdc45  
3 ENSMUSG000…  2.15e3   0.458 0.765  0.599  0.549 0.995 NS    H19    
4 ENSMUSG000…  2.49e1   0.579 0.561  1.03   0.302 0.951 NS    Scml2  
5 ENSMUSG000…  7.78e0  -0.899 1.55  -0.579  0.562 0.998 NS    Apoh   
6 ENSMUSG000…  1.97e4  -0.174 0.204 -0.855  0.392 0.982 NS    Narf   
# … with abbreviated variable names ¹​baseMean, ²​log2FoldChange,
#   ³​external_gene_name</code></pre>
<p>We can use some of the <code>tidyverse</code> functions we’ve
encountered previously to rename the <code>external_gene_name</code>
column to <code>symbol</code> and to move it into the second column
position? Hint: Because of the order of the packages we may have loaded,
we’ll use <code>dplyr::rename()</code> and <code>dplyr::select()</code>
instead of just the <code>select()</code> function. We can discuss this
in a moment.</p>
<pre class="r"><code>results_deficient_vs_control_annotated = results_deficient_vs_control_annotated %&gt;%
    dplyr::rename(&#39;symbol&#39; = &#39;external_gene_name&#39;) %&gt;%
    dplyr::select(id, symbol, everything())
results_deficient_vs_control_annotated</code></pre>
<pre><code># A tibble: 16,249 × 9
   id          symbol baseM…¹ log2F…² lfcSE   stat pvalue  padj call 
   &lt;chr&gt;       &lt;chr&gt;    &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;fct&gt;
 1 ENSMUSG000… Gnai3   1.49e3   0.298 0.210  1.42   0.157 0.869 NS   
 2 ENSMUSG000… Cdc45   1.75e3   0.226 0.177  1.28   0.200 0.903 NS   
 3 ENSMUSG000… H19     2.15e3   0.458 0.765  0.599  0.549 0.995 NS   
 4 ENSMUSG000… Scml2   2.49e1   0.579 0.561  1.03   0.302 0.951 NS   
 5 ENSMUSG000… Apoh    7.78e0  -0.899 1.55  -0.579  0.562 0.998 NS   
 6 ENSMUSG000… Narf    1.97e4  -0.174 0.204 -0.855  0.392 0.982 NS   
 7 ENSMUSG000… Klf6    9.51e2  -0.145 0.222 -0.652  0.515 0.992 NS   
 8 ENSMUSG000… Scmh1   7.27e2   0.265 0.275  0.964  0.335 0.966 NS   
 9 ENSMUSG000… Cox5a   1.43e3  -0.272 0.212 -1.28   0.199 0.903 NS   
10 ENSMUSG000… Tbx2    1.52e1   0.542 0.687  0.788  0.430 0.985 NS   
# … with 16,239 more rows, and abbreviated variable names ¹​baseMean,
#   ²​log2FoldChange</code></pre>
<p>And now we have our differential expression results annotated with
gene symbols, which can help in the interpretation of the results, and
can be used in downstream analysis such as functional analysis.</p>
</div>
<div id="outputting-results-to-file" class="section level1">
<h1>Outputting results to file</h1>
<p>A key aspect of our analysis is preserving the relevant datasets for
both our records and for downstream applications, such as functional
enrichments.</p>
<div id="de-results-table" class="section level2">
<h2>DE results table</h2>
<p>We’ll write out our DE results, now that we’ve added information to
the table to help us or our collaborators interpret the results.</p>
<pre class="r"><code>write.csv(results_deficient_vs_control,
          row.names = FALSE,
          na = &quot;.&quot;,
          file=&quot;outputs/tables/DE_results_deficient_vs_control.csv&quot;)

write.csv(results_deficient_vs_control_annotated,
          row.names = FALSE,
          file=&quot;outputs/tables/DE_results_deficient_vs_control_annotated.csv&quot;)</code></pre>
</div>
<div id="r-session-data" class="section level2">
<h2>R session data</h2>
<p>In addition to the individual RObj(s) we saved earlier, we can
capture a snapshot our entire session using the <code>save.image</code>
function. This can be loaded in the same manner as an individual
Robj.</p>
<p>First, we’ll save our session info so we can reference the packages
and versions used to generate these data.</p>
<pre class="r"><code>session_summary &lt;- sessionInfo()</code></pre>
<pre class="r"><code>save.image(file = &quot;outputs/Robjs/DE_iron.RData&quot;)</code></pre>
</div>
</div>
<div id="overall-takeaways" class="section level1">
<h1>Overall takeaways</h1>
<p>We’ve run through most of the building blocks needed to run a
differential expression analysis and hopefully built up a better
understanding of how differential expression comparisons work,
particularly how experimental design can impact our results.</p>
<p>What to consider moving forward:</p>
<ul>
<li>How can I control for technical variation in my experimental
design?</li>
<li>How much variation is expected with a treatment group?</li>
<li>What is my RNA quality, and how can that be optimized?</li>
<li>Are there quality concerns for my sequencing data?</li>
<li>What comparisons are relevant to my biological question?</li>
<li>Are there covariates that should be considered?</li>
<li>What will a differential expression analysis tell me?</li>
</ul>
<p><strong>Let’s pause here for general questions</strong></p>
<hr />
</div>
<div id="but-how-do-we-make-sense-of-large-numbers-of-de-genes"
class="section level1">
<h1>But how do we make sense of large numbers of DE genes?</h1>
<p>A way to determine possible <a
href="https://www.ebi.ac.uk/training-beta/online/courses/functional-genomics-ii-common-technologies-and-data-analysis-methods/biological-interpretation-of-gene-expression-data-2/">broader
biological interpretations</a> from the observed DE results, is
functional enrichments. There are many options, such as some included in
this <a
href="https://www.researchgate.net/post/How_can_I_analyze_a_set_of_DEGs_differentially_expressed_genes_to_obtain_information_from_them">discussion
thread</a>. Other common functional enrichments approaches are gene set
enrichment analysis, aka <a
href="http://software.broadinstitute.org/gsea/index.jsp">GSEA</a>,
Database for Annotation, Visualization and Integrated Discovery, aka <a
href="https://david.ncifcrf.gov/">DAVID</a>, <a
href="https://digitalinsights.qiagen.com/">Ingenity</a>, and <a
href="https://advaitabio.com/ipathwayguide/">iPathway Guide</a></p>
<p>The University of Michigan has license and support for additional
tools, such as Cytoscape, so we recommend reaching out to staff with <a
href="https://www.lib.umich.edu/locations-and-hours/taubman-health-sciences-library/research-and-clinical-support">Taubman
Library</a> to learn more about resources that might be application
toyour research.</p>
<hr />
</div>
<div id="sources" class="section level1">
<h1>Sources</h1>
<ul>
<li>HBC DGE training module, part 1: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html</a></li>
<li>HBC DGE training module, part 2: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html</a></li>
<li>DESeq2 vignette: <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis"
class="uri">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis</a></li>
<li>Bioconductor Genomic Annotation resources: <a
href="http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html"
class="uri">http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html</a></li>
<li>BioMart vignette: <a
href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html"
class="uri">https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html</a></li>
</ul>
</div>
<div id="additional-resources" class="section level1">
<h1>Additional Resources</h1>
<ul>
<li>MIDAS Reproduciblity Hub: <a
href="https://midas.umich.edu/reproducibility-overview/"
class="uri">https://midas.umich.edu/reproducibility-overview/</a></li>
<li>ARC resources: <a href="https://arc-ts.umich.edu/"
class="uri">https://arc-ts.umich.edu/</a></li>
<li>Gene Set Enrichment Resources from Bioconductor: <a
href="https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html"
class="uri">https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html</a></li>
<li>Using HTSeq data with DESeq2: <a
href="https://angus.readthedocs.io/en/2019/diff-ex-and-viz.html"
class="uri">https://angus.readthedocs.io/en/2019/diff-ex-and-viz.html</a></li>
<li>Detailed RNA-seq analysis paper: <a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6096346/"
class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6096346/</a></li>
<li>Overview of RNA-seq analysis considerations: <a
href="https://academic-oup-com.proxy.lib.umich.edu/bfg/article/14/2/130/257370"
class="uri">https://academic-oup-com.proxy.lib.umich.edu/bfg/article/14/2/130/257370</a></li>
<li>Alternative overview of DESeq2, including visualizations and
functional enrichments: <a
href="http://dputhier.github.io/jgb71e-polytech-bioinfo-app/practical/rna-seq_R/rnaseq_diff_Snf2.html"
class="uri">http://dputhier.github.io/jgb71e-polytech-bioinfo-app/practical/rna-seq_R/rnaseq_diff_Snf2.html</a></li>
</ul>
<hr />
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.0 (2022-04-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS 13.2.1

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets 
[7] methods   base     

other attached packages:
 [1] forcats_0.5.2               stringr_1.5.0              
 [3] purrr_1.0.1                 readr_2.1.3                
 [5] tibble_3.1.8                tidyverse_1.3.2            
 [7] biomaRt_2.52.0              data.table_1.14.6          
 [9] dplyr_1.0.10                knitr_1.41                 
[11] rmarkdown_2.18              RColorBrewer_1.1-3         
[13] pheatmap_1.0.12             ggrepel_0.9.2              
[15] tidyr_1.3.0                 ggplot2_3.4.0              
[17] DESeq2_1.36.0               SummarizedExperiment_1.26.1
[19] Biobase_2.56.0              MatrixGenerics_1.8.1       
[21] matrixStats_0.63.0          GenomicRanges_1.48.0       
[23] GenomeInfoDb_1.32.4         IRanges_2.30.1             
[25] S4Vectors_0.34.0            BiocGenerics_0.42.0        

loaded via a namespace (and not attached):
 [1] googledrive_2.0.0      colorspace_2.0-3      
 [3] ellipsis_0.3.2         XVector_0.36.0        
 [5] fs_1.5.2               rstudioapi_0.14       
 [7] farver_2.1.1           bit64_4.0.5           
 [9] AnnotationDbi_1.58.0   fansi_1.0.3           
[11] lubridate_1.9.0        xml2_1.3.3            
[13] codetools_0.2-18       splines_4.2.0         
[15] cachem_1.0.6           geneplotter_1.74.0    
[17] jsonlite_1.8.3         broom_1.0.1           
[19] annotate_1.74.0        dbplyr_2.2.1          
[21] png_0.1-8              compiler_4.2.0        
[23] httr_1.4.4             backports_1.4.1       
[25] assertthat_0.2.1       Matrix_1.5-3          
[27] fastmap_1.1.0          gargle_1.2.1          
[29] cli_3.4.1              htmltools_0.5.3       
[31] prettyunits_1.1.1      tools_4.2.0           
[33] gtable_0.3.1           glue_1.6.2            
[35] GenomeInfoDbData_1.2.8 rappdirs_0.3.3        
[37] Rcpp_1.0.9             cellranger_1.1.0      
[39] jquerylib_0.1.4        vctrs_0.5.2           
[41] Biostrings_2.64.1      xfun_0.35             
[43] rvest_1.0.3            timechange_0.1.1      
[45] lifecycle_1.0.3        XML_3.99-0.13         
[47] googlesheets4_1.0.1    zlibbioc_1.42.0       
[49] scales_1.2.1           vroom_1.6.0           
[51] hms_1.1.2              parallel_4.2.0        
[53] yaml_2.3.6             curl_4.3.3            
[55] memoise_2.0.1          sass_0.4.4            
[57] stringi_1.7.8          RSQLite_2.2.19        
[59] highr_0.9              genefilter_1.78.0     
[61] filelock_1.0.2         BiocParallel_1.30.4   
[63] rlang_1.0.6            pkgconfig_2.0.3       
[65] bitops_1.0-7           evaluate_0.18         
[67] lattice_0.20-45        labeling_0.4.2        
[69] bit_4.0.5              tidyselect_1.2.0      
[71] magrittr_2.0.3         R6_2.5.1              
[73] generics_0.1.3         DelayedArray_0.22.0   
[75] DBI_1.1.3              pillar_1.8.1          
[77] haven_2.5.1            withr_2.5.0           
[79] survival_3.4-0         KEGGREST_1.36.3       
[81] RCurl_1.98-1.9         modelr_0.1.10         
[83] crayon_1.5.2           utf8_1.2.2            
[85] BiocFileCache_2.4.0    tzdb_0.3.0            
[87] progress_1.2.2         locfit_1.5-9.6        
[89] grid_4.2.0             readxl_1.4.1          
[91] blob_1.2.3             reprex_2.0.2          
[93] digest_0.6.30          xtable_1.8-4          
[95] munsell_0.5.0          bslib_0.4.1           </code></pre>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="Module11_DEVisualizations.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="R_bonus_content.html">Next lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>

<div id="rmd-source-code">---
title: "DE Annotations & Downstream applications"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("11-")
```

In this module, we will learn:

* The advantages of using gene ids when analyzing RNA-seq data.
* How to add gene symbols and other annotations, using ENSEMBL gene ids
* Discuss options for functional enrichments and other follow-ups

<br>

```{r Modules, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
library(ggplot2)
library(tidyr)
library(dplyr)
library(matrixStats)
library(ggrepel)
library(pheatmap)
library(RColorBrewer)
# load("rdata/RunningData.RData")

## ensembl SSL cert issue flagged in knitting with selected mirror
```

# Differential Expression Workflow {.unlisted .unnumbered}

Here we will generate summary figures for our results and annotate our DE tables.

![](./images/wayfinder/wayfinder-DEVisualizations.png){width=75%}

---

# Adding genome annotations

Since, gene symbols can change over time or be ambiguous we use, and recommend, using the EMSEMBL reference genome and ENSEMBL IDs for alignments and we've been working with tables and data where all genes are labeled only by their long ENSEMBL ID. However, this can make it difficult to quickly look for genes of interest.

Luckily, Bioconductor provides many tools and resources to facilitate access to [genomic annotation resources](http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html).

To start, we will first load the [biomaRt library](https://bioconductor.org/packages/3.14/bioc/html/biomaRt.html) and choose what reference we want to access. For a more detailed walk through of using biomaRt, [this training module](https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2019/RNAseq/html/05_Annotation_and_Visualisation.html) might be useful, including what to do when annotations are not 1:1 mappings.

We'll start by loading the `biomaRt` library and calling the `useEnsembl()` function to select the database we'll use to extract the information we need. This will download the mapping of ENSEMBL IDs to gene symbols, enabling us to eventually add the gene symbol column we want.

```{r Pullmart, warning=FALSE}
library('biomaRt')
ensembl = useEnsembl(dataset = 'mmusculus_gene_ensembl', biomart='ensembl', mirror = "useast")
```

**Note** *- this process takes some time and will take up a larger amount of working memory so proceed with caution if you try to run these commands on a laptop with less than 4G of memory*

To identify possible **filters** to restrict our data, we can use the `listFilters` function. To identify the **attributes** we want to retrive, we can use the `listAttributes` function. The best approach is to use [list or search functions](https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html#how-to-build-a-biomart-query) to help narrow down the available options.

```{r AddAnnotations2, warning=FALSE, eval=FALSE}
head(listFilters(mart = ensembl), n = 20)
head(listAttributes(ensembl), n = 30)
```

We can access additional genomic annotations using the [`bioMart` package](https://bioconductor.org/packages/release/bioc/html/biomaRt.html). To identify  we'll structure our 'query' or search of the bioMart resources to use the [ENSEMBL id](https://m.ensembl.org/info/genome/genebuild/gene_names.html) from our alignment to add the gene symbols and gene description for each gene.

```{r AddAnotation3, warning=FALSE}
id_mapping = getBM(attributes=c('ensembl_gene_id', 'external_gene_name'),
      filters = 'ensembl_gene_id',
      values = row.names(assay(dds_fitted)),
      mart = ensembl)
# will take some time for query

# Preview the result
head(id_mapping)
```

Now that we have the ENSEMBL information and a gene symbol to match to our results, we can proceed in the smaller groups. As with the previous exercise, we have broken it into small steps with hints as needed.

**Note**: For additional information regarding bioMart, please consult the [ENSEMBL bioMart vignette](https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html) or the broader [Bioconductor Annotation Resources vignette ](http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html).

Look at the two data frames that are going to be needed: `id_mapping` and `results_deficient_vs_control`.

```{r preview_tables}
head(id_mapping)
head(results_deficient_vs_control)
```

We want to match the `id` column of `results_deficient_vs_control` to the `ensembl_gene_id` column of `id_mapping`, and once that match is found, we want to extract the `external_gene_name` column of `id_mapping` to get the gene symbol. Next, look at the documentation for `dplyr::left_join()` and merge the `id_mapping` table into the `results_deficient_vs_control` table on the columns `ensembl_gene_id` and `external_gene_name`.

```{r left_join}
results_deficient_vs_control_annotated = as_tibble(results_deficient_vs_control, rownames = "id") %>%
    left_join(id_mapping, by = c('id' = 'ensembl_gene_id'))
head(results_deficient_vs_control_annotated)
```

We can use some of the `tidyverse` functions we've encountered previously to rename the `external_gene_name` column to `symbol` and to move it into the second column position? Hint: Because of the order of the packages we may have loaded, we'll use `dplyr::rename()` and `dplyr::select()` instead of just the `select()` function. We can discuss this in a moment.

```{r rename_rearrange}
results_deficient_vs_control_annotated = results_deficient_vs_control_annotated %>%
    dplyr::rename('symbol' = 'external_gene_name') %>%
    dplyr::select(id, symbol, everything())
results_deficient_vs_control_annotated
```

And now we have our differential expression results annotated with gene symbols, which can help in the interpretation of the results, and can be used in downstream analysis such as functional analysis.

# Outputting results to file

A key aspect of our analysis is preserving the relevant datasets for both our records and for downstream applications, such as functional enrichments.

## DE results table

We'll write out our DE results, now that we've added information to the table to help us or our collaborators interpret the results.

```{r DEResultsOuput, eval = FALSE}
write.csv(results_deficient_vs_control,
          row.names = FALSE,
          na = ".",
          file="outputs/tables/DE_results_deficient_vs_control.csv")

write.csv(results_deficient_vs_control_annotated,
          row.names = FALSE,
          file="outputs/tables/DE_results_deficient_vs_control_annotated.csv")
```


## R session data

In addition to the individual RObj(s) we saved earlier, we can capture a snapshot our entire session using the `save.image` function. This can be loaded in the same manner as an individual Robj.

First, we'll save our session info so we can reference the packages and versions used to generate these data.

```{r}
session_summary <- sessionInfo()
```

```{r }
save.image(file = "outputs/Robjs/DE_iron.RData")
```



# Overall takeaways

We've run through most of the building blocks needed to run a differential expression analysis and hopefully built up a better understanding of how differential expression comparisons work, particularly how experimental design can impact our results.

What to consider moving forward:

* How can I control for technical variation in my experimental design?
* How much variation is expected with a treatment group?
* What is my RNA quality, and how can that be optimized?
* Are there quality concerns for my sequencing data?
* What comparisons are relevant to my biological question?
* Are there covariates that should be considered?
* What will a differential expression analysis tell me?


**Let's pause here for general questions**

---

# But how do we make sense of large numbers of DE genes?

A way to determine possible [broader biological interpretations](https://www.ebi.ac.uk/training-beta/online/courses/functional-genomics-ii-common-technologies-and-data-analysis-methods/biological-interpretation-of-gene-expression-data-2/) from the observed DE results, is functional enrichments. There are many options, such as some included in this [discussion thread](https://www.researchgate.net/post/How_can_I_analyze_a_set_of_DEGs_differentially_expressed_genes_to_obtain_information_from_them). Other common functional enrichments approaches are gene set enrichment analysis, aka [GSEA](http://software.broadinstitute.org/gsea/index.jsp), Database for Annotation, Visualization and Integrated Discovery, aka [DAVID](https://david.ncifcrf.gov/), [Ingenity](https://digitalinsights.qiagen.com/), and [iPathway Guide](https://advaitabio.com/ipathwayguide/)

The University of Michigan has license and support for additional tools, such as Cytoscape, so we recommend reaching out to staff with [Taubman Library](https://www.lib.umich.edu/locations-and-hours/taubman-health-sciences-library/research-and-clinical-support) to learn more about resources that might be application toyour research.

---

# Sources

* HBC DGE training module, part 1: https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html
* HBC DGE training module, part 2: https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html
* DESeq2 vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis
* Bioconductor Genomic Annotation resources: http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html
* BioMart vignette: https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html

# Additional Resources
* MIDAS Reproduciblity Hub: https://midas.umich.edu/reproducibility-overview/
* ARC resources: https://arc-ts.umich.edu/
* Gene Set Enrichment Resources from Bioconductor: https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html
* Using HTSeq data with DESeq2: https://angus.readthedocs.io/en/2019/diff-ex-and-viz.html
* Detailed RNA-seq analysis paper: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6096346/
* Overview of RNA-seq analysis considerations: https://academic-oup-com.proxy.lib.umich.edu/bfg/article/14/2/130/257370
* Alternative overview of DESeq2, including visualizations and functional enrichments: http://dputhier.github.io/jgb71e-polytech-bioinfo-app/practical/rna-seq_R/rnaseq_diff_Snf2.html

---

```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
# Hidden code block to write out data for knitting
# save.image(file = "rdata/RunningData_Full.RData")
```

# Session Info
```{r SessionInfo}
sessionInfo()
```

---

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](Module11_DEVisualizations.html) | [Top of this lesson](#top) | [Next lesson](R_bonus_content.html) |
| :--- | :----: | ---: |</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module12_DEAnnotations.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
